src/esp/assets/PTexMeshData.cpp

```diff
@@ -582,54 +660,75 @@ void PTexMeshData::uploadBuffersToGPU(bool forceReload) {
     currentMesh->ibo.setData(submeshes_[iMesh].ibo,
                              Magnum::GL::BufferUsage::StaticDraw);
   }
-  LOG(INFO) << "... done" << std::endl;
 
   LOG(INFO) << "Calculating mesh adjacency... ";
-
   std::vector<std::vector<uint32_t>> adjFaces(submeshes_.size());
 
+  // load it if it is computed before.
+  // otherwise compute it once and save it for future usage.
+  const std::string adjFaceFilename =
+      Corrade::Utility::Directory::join(atlasFolder_, "../adjFaces.bin");
+  if (!loadAdjacency(adjFaceFilename, adjFaces)) {
 #pragma omp parallel for
-  for (int iMesh = 0; iMesh < submeshes_.size(); ++iMesh) {
-    calculateAdjacency(submeshes_[iMesh], adjFaces[iMesh]);
+    for (int iMesh = 0; iMesh < submeshes_.size(); ++iMesh) {
+      calculateAdjacency(submeshes_[iMesh], adjFaces[iMesh]);
+    }
+    // Warning: you should have enough disk space to store the info
+    // it usually takes a couple of 100MB (usually 200+MB).
+    saveAdjacency(adjFaceFilename, adjFaces);
+    LOG(INFO) << "Done: it is computed and saved to: " << adjFaceFilename;
+  } else {
+    LOG(INFO) << "Done: Loaded it from " << adjFaceFilename;
   }
 
   for (int iMesh = 0; iMesh < submeshes_.size(); ++iMesh) {
     auto& currentMesh = renderingBuffers_[iMesh];
 
-    currentMesh->adjTex.setBuffer(Magnum::GL::BufferTextureFormat::R32UI,
-                                  currentMesh->abo);
+    currentMesh->adjFaces.setBuffer(Magnum::GL::BufferTextureFormat::R32UI,
+                                    currentMesh->abo);
     currentMesh->abo.setData(adjFaces[iMesh],
                              Magnum::GL::BufferUsage::StaticDraw);
+
+    // experiment code (may not work):
+    // using GL_LINES_ADJACENCY here to send quads to geometry shader
```

Comment:
@mosra: I believe it is the most suspicious part.
The purpose is to send quads to geometry shader using GL_LINES_ADJACENCY.
The corresponding code in RelicaSDK is:
`
glDrawElements(GL_LINES_ADJACENCY, mesh.ibo.num_elements, mesh.ibo.datatype, 0);
`
Did I set it correctly using magnum?

Reply:
Yes, I think so (does it work?).